### PR TITLE
feat(openid-connect): allow set headers in introspection request

### LINKED ISSUE #10601: help request: How to set header to introspection endpoint used within OIDC plugin?
### Description

In other API gateway I used similar way to set header to my introspection endpoint that I used within OIDC plugin. In pre-function plugin I did set header using their code `gateway.service.request.set_header('header', 'value')`. However, with APISIX, I'm unable to replicate this. Even though I added `serverless-pre-function` plugin with phase set to `rewrite` or `access`, with `ngx.req.set_header` method, header was not set and forwarded to introspection endpoint.

### Environment

- APISIX version: 3.7
- Operating system: Linux apisix-dev-7b89dc65c5-54f8w 5.10.176-157.645.amzn2.x86_64 #1 SMP Tue Mar 28 17:49:06 UTC 2023 x86_64 GNU/Linux
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): openresty/1.21.4.2 / 
- Plugin runner version, for issues related to plugin runners: no sure what is this and how to get it
- LuaRocks version, for installation issues (run `luarocks --version`): no luarocks installed

### PR DESCRIPTION
**Pull request by yuweizzz**
### Description

Fixes #10601 

allow set headers in introspection request.

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by shreemaan-abhishek** [comment]
could you please explain about the use cases where this feature would be helpful in the linked issue thread?

**Comment by yuweizzz** [comment]
this pr allows provide extra HTTP headers when talks to the OpenID Connect introspection endpoint, In other API gateway has this feature alrealy, like traefik. and some oauth libs has this feature too, like oauthlib. this may helpful when introspection endpoint is proxied.

**Comment by shreemaan-abhishek** [comment]
@yuweizzz please fix the linter

**Comment by yuweizzz** [comment]
done.

File: apisix/plugins/openid-connect.lua [hunk 1]
```diff
@@ -386,7 +400,20 @@ local function introspect(ctx, conf)
     else
         -- Validate token against introspection endpoint.
         -- TODO: Same as above for public key validation.
+        if conf.introspection_addon_headers then
+            -- http_request_decorator option provides by lua-resty-openidc
```
Comment (shreemaan-abhishek) [inline apisix/plugins/openid-connect.lua:404]:
```suggestion
            -- http_request_decorator option provided by lua-resty-openidc
```

File: t/plugin/openid-connect6.t [hunk 2]
```diff
@@ -155,3 +155,213 @@ passed
...
+                                "discovery": "http://127.0.0.1:8080/realms/University/.well-known/openid-configuration",
+                                "redirect_uri": "http://localhost:3000",
+                                "ssl_verify": false,
+                                "timeout": 10,
+                                "bearer_only": true,
+                                "realm": "University",
+                                "introspection_endpoint_auth_method": "client_secret_post",
+                                "introspection_endpoint": "http://127.0.0.1:8080/realms/University/protocol/openid-connect/token/introspect",
+                                "introspection_addon_headers": {
+                                    "X-Addon-Header-A": "VALUE",
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect6.t:182]: hardcoding values in configuration doesn't make sense, I think the configuration should just specify header fields which would be extracted from the original request.

Comment (yuweizzz) [reply t/plugin/openid-connect6.t:182]: do you mean change the headers from the original request?

Comment (shreemaan-abhishek) [reply t/plugin/openid-connect6.t:182]: no, the plugin configuration should only specify which headers should be sent. APISIX should extract those headers from the request and then add them to the introspection request.

Comment (yuweizzz) [reply t/plugin/openid-connect6.t:182]: I thinking it's hard to decide which headers should be sent, maybe we should just keep it open here and leave it to user customization?

File: t/plugin/openid-connect6.t [hunk 3]
```diff
@@ -155,3 +155,213 @@ passed
...
+                local body = json_decode(res.body)
+                local accessToken = body["access_token"]
+
+                -- Access route using access token. Should work.
+                uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/hello"
+                local res, err = httpc:request_uri(uri, {
+                    method = "GET",
+                    headers = {
+                        ["Authorization"] = "Bearer " .. body["access_token"],
+                        ["X-Addon-Header-A"] = "Value-A",
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect6.t:240]: what happens if we don't pass these headers?

File: t/plugin/openid-connect6.t [hunk 4]
```diff
@@ -155,3 +155,213 @@ passed
...
+                ngx.say(false)
+            end
+        }
+    }
+--- response_body
+true
+--- grep_error_log eval
+qr/token validate successfully by \w+/
+--- grep_error_log_out
+token validate successfully by introspection
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect6.t:263]: I think we can check this just like so:

```perl
--- error_log
token validate successfully by introspection
```

File: t/plugin/openid-connect6.t [hunk 5]
```diff
@@ -155,3 +155,213 @@ passed
...
+
+
+
+=== TEST 6: Access route with an invalid token, should work as expected too.
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect6.t:267]:
```suggestion
=== TEST 6: Access route with an invalid token, should fail
```

File: t/plugin/openid-connect6.t [hunk 6]
```diff
@@ -155,3 +155,213 @@ passed
...
+            end
+            ngx.say(body)
+        }
+    }
+--- response_body
+passed
+
+
+
+=== TEST 8: Check http headers from fake introspection endpoint.
```
Comment (shreemaan-abhishek) [inline t/plugin/openid-connect6.t:342]: I don't understand what's happening in this test.

File: docs/en/latest/plugins/openid-connect.md [hunk 7]
```diff
@@ -89,6 +89,7 @@ description: OpenID Connect allows the client to obtain user information from th
 | cache_segment | string | False |  |  | Optional name of a cache segment, used to separate and differentiate caches used by token introspection or JWT verification. |
 | introspection_interval | integer | False | 0 |  | TTL of the cached and introspected access token in seconds. |
 | introspection_expiry_claim | string | False |  |  | Name of the expiry claim, which controls the TTL of the cached and introspected access token. The default value is 0, which means this option is not used and the plugin defaults to use the TTL passed by expiry claim defined in `introspection_expiry_claim`. If `introspection_interval` is larger than 0 and less than the TTL passed by expiry claim defined in `introspection_expiry_claim`, use `introspection_interval`. |
+| introspection_addon_headers | string[] | False |  |  | Array of strings. Used in Append extras headers to the introspection http request, If configured header not exist in origin request, it will be ignore. |
```
Comment (kayx23) [inline docs/en/latest/plugins/openid-connect.md:92]:
```suggestion
| introspection_addon_headers | string[] | False |  |  | Array of strings. Used to append additional header values to the introspection HTTP request. If the specified header does not exist in origin request, value will not be appended. |
```

File: t/plugin/openid-connect6.t (hunk 3 above)
Comment (yuweizzz) [reply t/plugin/openid-connect6.t:240]: the test just for compare with test 6 and test 8. It should work normally if we don't pass these headers.

File: t/plugin/openid-connect6.t (hunk 6 above)
Comment (yuweizzz) [reply t/plugin/openid-connect6.t:342]: test 8 make a fake introspection endpoint which logs requests header, so we can check if those header same as the origin request.

**Comment by yuweizzz** [comment]
any update?

**Review by shreemaan-abhishek — APPROVED** [review]

**Review by starsz — APPROVED** [review]

**Pull request merged**